hollow/src/main/java/com/netflix/hollow/api/consumer/index/PrimaryKeyIndex.java

```diff
@@ -0,0 +1,319 @@
+     * @param primaryType the primary type
+     * @param <T> the primary type
+     * @return a builder
+     */
+    public static <T extends HollowObject> Builder<T> from(HollowConsumer consumer, Class<T> primaryType) {
+        Objects.requireNonNull(consumer);
+        Objects.requireNonNull(primaryType);
+        return new Builder<>(consumer, primaryType);
+    }
+
+    /**
+     * The builder of a {@link PrimaryKeyIndex}.
+     * <p>
+     * The default configuration specifies, unless specified differently by an initiating builder method, that
+     * the hash index listen to {@code HollowConsumer} version updates as if by a call to {@link #listenToDataRefresh}
+     * with a value of {@code true}.
+     *
+     * @param <T> the primary type
+     */
+    public static final class Builder<T extends HollowObject> {
+        final HollowConsumer consumer;
+        final Class<T> primaryType;
+        PrimaryKey primaryTypeKey; // non-null on bindWithPrimaryKeyOnType
+        boolean listenToDataRefresh = true;
+
+        Builder(HollowConsumer consumer, Class<T> primaryType) {
+            this.consumer = consumer;
+            this.primaryType = primaryType;
+        }
+
+        /**
+         * Binds the field paths with those of the primary key associated with the schema of the primary type.
+         *
+         * @throws com.netflix.hollow.api.error.SchemaNotFoundException if there is no schema for the primary
+         * type
+         * @throws IllegalArgumentException if there is no primary key associated with the primary type
+         */
+        public Builder<T> bindWithPrimaryKeyOnType() {
+            String primaryTypeName = primaryType.getSimpleName();
+            HollowSchema schema = consumer.getStateEngine().getNonNullSchema(primaryTypeName);
+
+            assert schema.getSchemaType() == HollowSchema.SchemaType.OBJECT;
+
+            this.primaryTypeKey = ((HollowObjectSchema) schema).getPrimaryKey();
+            if (primaryTypeKey == null) {
+                throw new IllegalArgumentException(
+                        String.format("No primary key associated with primary type %s", primaryType));
+            }
+            return this;
+        }
+
+        /**
+         * Configures the primary key index to listen to {@code HollowConsumer} version updates.
+         * On an update the index recalculates so updated data will be reflected in the results of a query
+         * (performed after the update).
+         *
+         * @param listen {@code true} if the index listens to {@code HollowConsumer} version updates,
+         * otherwise {@code false} and the index does not listen to {@code HollowConsumer} updates.  The default
+         * value is {@code true}, unless specified differently by the initiating builder method returning this builder.
+         * @return this builder
+         */
+        public Builder<T> listenToDataRefresh(boolean listen) {
```

Comment:
I wonder if there's a more conventional name to use here that we can switch to since we don't have to worry about backwards compatibility?

Or flip it around and make `PrimaryKeyIndex` implement `HollowConsumer.RefreshListener` and have callers call `consumer.addRefreshListener(myIndex)` explicitly?